GPFlow 2.0: ARD and Stationaries Refactor

jahall opened the pull request:
Migrating the generalized periodic enhancement from 1.0. This is the initial groundwork to make that happen i.e. addresses the issues outlined in https://github.com/GPflow/GPflow/issues/1077. In addition some minor refactoring has been done e.g. to move Periodic out of misc.py and into periodic.py.

Will re-introduce the actual enhancement in a follow up commit tomorrow.

jahall commented:
@st-- Would welcome your feedback on the initial groundwork.

jahall commented:
@st-- May need your help tracking down the bugs thrown up in CI. I think they are related to my removal of the `ard` kwarg and associated ard validation check.

st-- commented on gpflow/kernels/base.py:
```diff
@@ -117,6 +120,19 @@ def slice_cov(self, cov: tf.Tensor) -> tf.Tensor:
+        X_shape = tf.shape(X)
+        l_shape = tf.shape(parameter)
+        if l_shape[-1] != X_shape[-1]:
+            # TODO: Maybe this should be a tf.errors.InvalidArgumentError
```
I don't think we should use another package's exceptions, especially if they're that unspecific...

jahall replied:
Agreed - will remove comment

st-- commented:
@jahall thanks for your work on this. I think (at least some of) the tests fail because the `ard` flag is actually used in the kernel expectations (`expectations/`). We could keep an `ard` property that checks whether the lengthscale parameter is a scalar (ard=False) or a vector (ard=True)...

I'm wondering if it'd be cleaner (& easier to review) to move the refactoring of ARD and `K_r2`/`K_r` into its own PR?

st-- commented on gpflow/kernels/linears.py:
```diff
@@ -13,24 +13,30 @@ class Linear(Kernel):
         - input_dim is the dimension of the input to the kernel
         - variance is the (initial) value for the variance parameter(s)
-          if ard=True, there is one variance per input
+          defaults to 1.0 but supports ARD i.e. one variance per input.
```
Could we have a more complete explanation? e.g. if you want ARD but initialised to 1 you'd have to pass in [1, 1, ..., 1] (where the length of the array is equal to the input dimension)

st-- replied:
(Applies to all kernels with ard feature)

jahall replied:
Makes sense - have done this

st-- commented on gpflow/kernels/base.py:
```diff
@@ -117,6 +120,24 @@ def slice_cov(self, cov: tf.Tensor) -> tf.Tensor:
 
         return cov
 
+    def validate_ard_parameter(self, X, parameter):
```
Could also check whether `active_dims` has the right length (if not None)?

jahall replied:
I will add in a check on __init__ to ensure size of `active_dims` has the right length (if not None and if not a slice)

st-- commented on gpflow/kernels/stationaries.py:
```diff
@@ -17,48 +17,54 @@ class Stationary(Kernel):
         - variance is the (initial) value for the variance parameter
         - lengthscale is the initial value for the lengthscale parameter
-          defaults to 1.0 (ard=False) or np.ones(input_dim) (ard=True).
+          defaults to 1.0.
```
explain how to get ard behaviour?

st-- commented on gpflow/kernels/stationaries.py:
```diff
@@ -17,48 +17,54 @@ class Stationary(Kernel):
         """
-        Returns the kernel evaluated on `r`, which is the scaled Euclidean distance
-        Should operate element-wise on r
+        Returns the kernel evaluated on `r²`, which is the squared scaled Euclidean distance
```
```suggestion
        Returns the kernel evaluated on r² (`r2`), which is the squared scaled Euclidean distance
```

st-- commented on tests/test_features.py:
```diff
@@ -30,9 +30,8 @@ def test_inducing_points_inducing_variable_len(N, D):
-                                      lengthscale=np.random.uniform(0.5, 3., 5),
-                                      ard=True),
-    gpflow.kernels.Periodic(period=0.4, variance=1.8)
+                                      lengthscale=np.random.uniform(0.5, 3., 5)),
```
should the lengthscale be an array to ensure it's actually testing ard=True ?

st-- commented on tests/test_kernels.py:
```diff
@@ -137,15 +132,14 @@ def test_arccosine_non_implemented_order(order):
         gpflow.kernels.ArcCosine(order=order)
 
 
-@pytest.mark.parametrize('ard', [True, False])
```
I would keep this in - maybe as
```python
@pytest.mark.parametrize('weight_variances', [
    1.,  # ard = False
    [1., 1., 1.],  # ard = True
])
```
(and then take `weight_variances` out of the following parametrization)
(assumes input_dim=3)

st-- replied:
or did this already get tested elsewhere ? :thinking:

jahall replied:
Actually - this is now sufficiently tested in `test_arccosine_kern_err` - this test can be removed

st-- commented on tests/test_kernels.py:
```diff
@@ -165,8 +159,6 @@ def _assert_periodic_kern_err(lengthscale, variance, period, X):
     gram_matrix = kernel(X)
     reference_gram_matrix = _ref_periodic(X, lengthscale, variance, period)
 
-    assert_allclose(gram_matrix, reference_gram_matrix)
```
why did this get removed?

st-- commented on tests/test_kernels.py:
```diff
@@ -243,7 +235,7 @@ def test_coregion_slice(N, input_dim, output_dim, rank):
     SquaredExponential() + Linear(),
     SquaredExponential() * Linear(),
-    SquaredExponential() + Linear(ard=True, variance=rng.rand(_dim, 1).reshape(-1))
+    SquaredExponential() + Linear(variance=rng.rand(_dim, 1).reshape(-1))
```
Why not
```suggestion
    SquaredExponential() + Linear(variance=rng.rand(_dim))
```
?

jahall replied:
Ha - yeah I'm not sure why that was defined that way. I'll change it.

st-- commented on tests/test_kernels.py:
```diff
@@ -362,7 +354,7 @@ def test_active_product(N, D):
         np.hstack([ls[:rand_idx], ls[rand_idx + 1:]]), ls[rand_idx], ls
     ]
     kernels = [
-        gpflow.kernels.SquaredExponential(lengthscale=lengthscale, active_dims=dims, ard=True)
```
does ard=True still get tested?

st-- commented on tests/test_kernels.py:
```diff
@@ -389,23 +381,9 @@ def test_ard_init_scalar(D):
-
-
-@pytest.mark.parametrize('D', [4, 7])
-def test_ard_init_MLP(D):
```
why did this disappear?

jahall replied:
This disappeared because the test is no longer valid - it was checking that if you passed in a scalar `weight_variances` with `ard=True` that it got expanded to an array. We now rely on the user explicitly specifying an array.

jahall commented:
@st-- Have addressed all comments. However, I cannot for the life of me figure out how to get this `validate_ard_parameter` method to work properly as `tf.cond` doesn't always lazily evaluate :( - see TODO in code. Other than that I think this is in reasonable shape. Just fixing the notebooks.

awav commented on gpflow/kernels/base.py:
```diff
@@ -117,6 +121,42 @@ def slice_cov(self, cov: tf.Tensor) -> tf.Tensor:
+                             f"dimension of the ard parameter ({len(raw_parameter)})")
+        return is_ard
+
+    def validate_ard_parameter(self, X, parameter):
```
This is wierd function, it just returns True? Why did you add this TODO and commented out code?

jahall replied:
I have removed this altogether for now as I cannot figure out how to perform this validation correctly during graph execution.

awav commented on tests/test_kernels.py:
```diff
@@ -380,32 +364,38 @@ def test_ard_init_scalar(D):
+        gpflow.kernels.SquaredExponential(lengthscale=np.ones(2), active_dims=[1])
+
+
+#@pytest.mark.parametrize('N', [4, 7])
```
Why this part is commented out?

jahall replied:
This is the associated test for the aforementioned commented code.

st-- commented on gpflow/kernels/linears.py:
```diff
@@ -13,24 +13,23 @@ class Linear(Kernel):
-        # variance, self.ard = self._validate_ard_shape("variance", variance, ard)
-        self.ard = ard
         self.variance = Parameter(variance, transform=positive())
+        self.ard = self._is_ard(variance)
```
I think it's better as a property (as you had before), e.g. if someone decides to replace the `variance` attribute with a different Parameter object. Same applies to the other kernels as well of course!

jahall replied:
Good point - I have reintroduced `ard` as a parameter and improved the implementation by creating an `ARDKernel` base class to enable better code re-use.

awav commented:
Hello @jahall , I apologize for closing your PR. It was closed automatically when I moved `develop-2.0` to the `develop`. Could you re-open the PR? Thanks!

awav commented:
I think it is a bit of overkill to create a separate class because of the ard property of parameters (not even all of them) in a kernel :)

jahall commented:
> I think it is a bit of overkill to create a separate class because of the ard property of parameters (not even all of them) in a kernel :)

That's reasonable. I'll refactor to make the change less intrusive.

awav commented on gpflow/kernels/linears.py:
```diff
@@ -1,31 +1,36 @@
+        """
+        Whether ARD behaviour is active.
+        """
+        return len(self.variance.shape) > 0
```
self.variance.shape.ndims
```suggestion
        return self.variance.shape.ndims > 0
```

awav commented on gpflow/kernels/misc.py:
```diff
@@ -55,10 +51,16 @@ def __init__(self,
+        """
+        Whether ARD behaviour is active.
+        """
+        return len(self.weight_variances.shape) > 0
```
```suggestion
        return self.weight_variances.shape.ndims > 0
```

jahall commented:
@awav @st-- Any further changes required for this PR? I just ask as its been open over a month and contains necessary changes (i.e. the `K_r`/`K_r2` refactor in `stationaries`) to merge the generalized periodic kernel from v1 (see #1052)
